src/dataflow/src/sink/kafka.rs

```diff
@@ -176,105 +183,122 @@ pub fn kafka<G>(
         &name.clone(),
         |info| {
             let activator = stream.scope().activator_for(&info.address[..]);
-            move |input| {
-                // Grab all of the available Rows and put them in a queue before we
-                // send it over to Kafka. We Even though we want to do bounded work
-                // per sink invocation, we still need to remember all inputs as we
-                // receive them
-                input.for_each(|_, rows| {
-                    rows.swap(&mut vector);
+            let shutdown_button_ref = &mut shutdown_button;
+            *shutdown_button_ref = Some(ShutdownButton::new(
+                producer.clone(),
+                stream.scope().activator_for(&info.address[..]),
+            ));
 
-                    for (row, _time, diff) in vector.drain(..) {
-                        queue.push_back((row, diff));
-                    }
-                });
+            (
+                move |input| {
+                    let producer = producer.borrow();
+                    if let Some(ref producer) = *producer {
```

Comment:
FWIW I consider this alternative formulation to be pretty idiomatic:

```rust
if let Some(producer) = &*producer.borrow() {
```

Reply:
I've seen this reformulation a bunch of times and I am not sure I grok it - to check my understanding:
is the right way to read this from right to left

1. the `borrow()` gives us a reference to the object inside the refcell
2. the `*` dereferences that reference (giving us an option)
3. the `&` takes a reference to whats inside the option? or does it just take a reference to the option again?

the last bit is the most confusing one